kvbc/test/v4blockchain/v4_keys_history_test.cpp

```diff
@@ -0,0 +1,446 @@
+  merkle_updates.addUpdate(key1, "initial_val");
+  updates.add(merkle_category, std::move(merkle_updates));
+  auto wb = db->getBatch();
+  keys_history.addBlockKeys(updates, num_blocks + 1, wb);
+  db->write(std::move(wb));
+
+  // make sure that key1's history is correct
+  opt_val = keys_history.getVersionFromHistory(merkle_category, key1, num_blocks);
+  ASSERT_EQ(opt_val, std::nullopt);
+  opt_val = keys_history.getVersionFromHistory(merkle_category, key1, num_blocks + 1);
+  ASSERT_NE(opt_val, std::nullopt);
+  ASSERT_EQ(opt_val.value(), num_blocks + 1);
+
+  // add some additional blocks [num_blocks + 2, additional_blocks]
+  const BlockId additional_blocks = num_blocks + (BlockId)dist(gen);  // generate number
+  for (BlockId i = num_blocks + 2; i <= additional_blocks; i++) {
+    add_merkle_keys_to_block(keys_history, i, num_keys);
+    // key1 was last updated in block [num_blocks + 1]
+    opt_val = keys_history.getVersionFromHistory(merkle_category, key1, i);
+    ASSERT_NE(opt_val, std::nullopt);
+    ASSERT_EQ(opt_val.value(), num_blocks + 1);
+  }
+
+  // update key1 to block [additional_blocks + 1]
+  categorization::Updates updates1;
+  categorization::BlockMerkleUpdates merkle_updates1;
+  merkle_updates1.addUpdate(key1, "updated_val");
+  updates1.add(merkle_category, std::move(merkle_updates1));
+  wb = db->getBatch();
+  keys_history.addBlockKeys(updates1, additional_blocks + 1, wb);
+  db->write(std::move(wb));
+
+  // make sure that key1's history is correct
+  opt_val = keys_history.getVersionFromHistory(merkle_category, key1, additional_blocks);
+  ASSERT_NE(opt_val, std::nullopt);
+  ASSERT_EQ(opt_val.value(), num_blocks + 1);
+  opt_val = keys_history.getVersionFromHistory(merkle_category, key1, additional_blocks + 1);
+  ASSERT_NE(opt_val, std::nullopt);
+  ASSERT_EQ(opt_val.value(), additional_blocks + 1);
+
+  // add some more blocks [additional_blocks + 2, more_blocks]
+  const BlockId more_blocks = additional_blocks + (BlockId)dist(gen);  // generate number
+  for (BlockId i = additional_blocks + 2; i <= more_blocks; i++) {
+    add_merkle_keys_to_block(keys_history, i, num_keys);
+    // key1 was last updated on num_blocks + 1
+    opt_val = keys_history.getVersionFromHistory(merkle_category, key1, i);
+    ASSERT_NE(opt_val, std::nullopt);
+    ASSERT_EQ(opt_val.value(), additional_blocks + 1);
+  }
+
+  // make sure also after we are done adding blocks
+  const BlockId total_blocks_number = num_blocks + additional_blocks + more_blocks;
+  for (BlockId i = 1; i <= total_blocks_number; i++) {
+    opt_val = keys_history.getVersionFromHistory(merkle_category, key1, i);
+    if (i <= num_blocks) {
+      ASSERT_EQ(opt_val, std::nullopt);
+    } else if (i <= additional_blocks) {
+      ASSERT_NE(opt_val, std::nullopt);
+      ASSERT_EQ(opt_val.value(), num_blocks + 1);
+    } else if (i <= more_blocks) {
+      ASSERT_NE(opt_val, std::nullopt);
+      ASSERT_EQ(opt_val.value(), additional_blocks + 1);
```

Comment:
To complete the sequence can you add also a deletion of key1 at the end with deletion version assertion as well, thanks

Reply:
great, done